pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -0,0 +1,498 @@
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).
+				Resource(&apiResource, v1.NamespaceAll).
+				Watch(&options)
+		},
+	}
+}
+
+func (gb *GraphBuilder) controllerFor(resource schema.GroupVersionResource, kind schema.GroupVersionKind) (*cache.Controller, error) {
+	// TODO: consider store in one storage.
+	glog.V(5).Infof("create storage for resource %s", resource)
+	client, err := gb.metaOnlyClientPool.ClientForGroupVersionKind(kind)
+	if err != nil {
+		return nil, err
+	}
+	gb.registeredRateLimiterForControllers.registerIfNotPresent(resource.GroupVersion(), client, "garbage_collector_monitoring")
+	setObjectTypeMeta := func(obj interface{}) {
+		runtimeObject, ok := obj.(runtime.Object)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("expected runtime.Object, got %#v", obj))
+		}
+		runtimeObject.GetObjectKind().SetGroupVersionKind(kind)
+	}
+	_, monitor := cache.NewInformer(
+		listWatcher(client, resource),
+		nil,
+		ResourceResyncTime,
+		cache.ResourceEventHandlerFuncs{
+			// add the event to the dependencyGraphBuilder's graphChanges.
+			AddFunc: func(obj interface{}) {
+				setObjectTypeMeta(obj)
+				event := &event{
+					eventType: addEvent,
+					obj:       obj,
+				}
+				gb.graphChanges.Add(event)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				setObjectTypeMeta(newObj)
+				// TODO: check if there are differences in the ownerRefs,
+				// finalizers, and DeletionTimestamp; if not, ignore the update.
+				event := &event{updateEvent, newObj, oldObj}
+				gb.graphChanges.Add(event)
+			},
+			DeleteFunc: func(obj interface{}) {
+				// delta fifo may wrap the object in a cache.DeletedFinalStateUnknown, unwrap it
+				if deletedFinalStateUnknown, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+					obj = deletedFinalStateUnknown.Obj
+				}
+				setObjectTypeMeta(obj)
+				event := &event{
+					eventType: deleteEvent,
+					obj:       obj,
+				}
+				gb.graphChanges.Add(event)
+			},
+		},
+	)
+	return monitor, nil
+}
+
+func (gb *GraphBuilder) monitorsForResources(resources map[schema.GroupVersionResource]struct{}) error {
+	for resource := range resources {
+		if _, ok := ignoredResources[resource]; ok {
+			glog.V(5).Infof("ignore resource %#v", resource)
+			continue
+		}
+		kind, err := gb.restMapper.KindFor(resource)
+		if err != nil {
+			return err
+		}
+		monitor, err := gb.controllerFor(resource, kind)
+		if err != nil {
+			return err
+		}
+		gb.monitors = append(gb.monitors, monitor)
+	}
+	return nil
+}
+
+func (gb *GraphBuilder) HasSynced() bool {
+	for _, monitor := range gb.monitors {
+		if !monitor.HasSynced() {
+			return false
+		}
+	}
+	return true
+}
+
+func (gb *GraphBuilder) Run(stopCh <-chan struct{}) {
+	for _, monitor := range gb.monitors {
+		go monitor.Run(stopCh)
+	}
+	go wait.Until(gb.runProcessGraphChanges, 1*time.Second, stopCh)
+}
+
+var ignoredResources = map[schema.GroupVersionResource]struct{}{
```

Comment:
"TODO: GC should detect if a resource requires its attention via API discovery."